### PR TITLE
Add initial governance vote analysis and CSV report scraping scripts

### PR DESCRIPTION
**Pull request by craigtutterow**
snapshot_report_scraper.py: Uses Selenium to emulate clicks to download CSV reports of governance votes based on provided list of url.

vote_analysis.ipynb: Jupyter notebook containing initial analysis and charts

vote_analysis.py: python executable for analysis code (requires parameter changes for running in local environment)

### CODE REVIEW

File: governance/snapshot-vote-analysis/snapshot_report_scraper.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
+from selenium import webdriver
+from selenium.webdriver.common.keys import Keys
+
+"""
+Prereqs:
+    - Install selenium python package (pip install selenium)
+    - Download chromedriver version matching your chrome browser: https://chromedriver.chromium.org/downloads
+"""
+
+vote_urls = {
```
Comment (josephd8) [inline governance/snapshot-vote-analysis/snapshot_report_scraper.py:10]: I had to remove the keys from this dictionary in order for this to work. It was trying to read the key value as the url.

File: governance/snapshot-vote-analysis/vote_analysis.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    'MVI': 'https://snapshot.org/#/index/proposal/QmadsabYMJC96jU2S2kPCSh1suVfDVApGLwrux2WwsHd7x',
+    'FLI2': 'https://snapshot.org/#/index/proposal/QmYHV2vdTaSubtNJefSoYx82ypsmtzrT7CGUU1EHsXzHC3',
+    'FLI1': 'https://snapshot.org/#/index/proposal/QmQwQn4k324kMKPjsSX6ZEzjkkKWh1DNfAN2mQ3dd5aP1a'
+}
+proposal_ids = [url.split('/')[-1] for url in vote_urls.values()]
+local_download_folder_path = 'C:/Users/craig/Downloads/'
+etherscan_token_hodler_csv_path = 'etherscan_index_hodler_table.csv'
+# Remove treasury, vesting and DEX LP wallets. Manually entered from https://etherscan.io/token/0x0954906da0Bf32d5479e25f46056d22f08464cab#balances
+wallet_addresses_to_exclude = [
+    '0xd89c642e52bd9c72bcc0778bcf4de307cc48e75a',
```
Comment (josephd8) [inline governance/snapshot-vote-analysis/vote_analysis.py:31]: We should also add "0x673d140eed36385cb784e279f8759f495c97cf03" to the list of wallet addresses to exclude for now. I am trying to track down what this contract is, but I am pretty sure it doesn't have voting rights.

File: governance/snapshot-vote-analysis/vote_analysis.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+circulating_supply = coin_gecko_response['market_data']['circulating_supply']
+quorum_threshold = 0.15
+yes_vote_threshold = 0.6
+votes_needed_for_quorum = round(project_quorum_threshold * circulating_supply, 2)
+etherscan_token_hodler_csv_path = 'etherscan_index_hodler_table.csv'
+
+def compute_gini(series_label:str, series_np_array:np.array):
+    gini = qe.gini_coefficient(series_np_array)
+    print(f'{series_label} Gini Coefficient: {gini}')
+    print(f'Global Wealth Gini Coefficient (2019): 0.885, source: https://en.wikipedia.org/wiki/List_of_countries_by_wealth_inequality')
```
Comment (josephd8) [inline governance/snapshot-vote-analysis/vote_analysis.py:62]: remove this line

File: governance/snapshot-vote-analysis/vote_analysis.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    votes_for = round(vote_df.loc[vote_df['choice'] == 1, 'balance'].sum(), 2)
+    voters_for = round(vote_df.loc[vote_df['choice'] == 1, 'choice'].sum(), 2)
+    votes_against = round(vote_df.loc[vote_df['choice'] == 2, 'balance'].sum(), 2)
+    voters_against = len(vote_df.loc[vote_df['choice'] == 2, 'choice'])
+    print(f'Votes for: {votes_for}, Votes against: {votes_against}')
+    print(f'Unique voters (wallets) for: {voters_for}, Unique voters (wallets) against: {voters_against}')
+    print(f'Did the vote exceed the threshold? {"Yes" if votes_for > votes_needed_for_quorum else "No"}')
+    if votes_for > votes_needed_for_quorum and votes_for/(votes_for + votes_against) > 0.6:
+        print(f'This initiative succeeded. {votes_for} votes in favor > {votes_needed_for_quorum} votes needed to meet quorum')
+        minimum_number_of_voters_needed = min(vote_df.index[vote_df['cumulative_dist'] > votes_needed_to_pass]) + 1
```
Comment (josephd8) [inline governance/snapshot-vote-analysis/vote_analysis.py:100]: `votes_needed_to_pass` is not a known variable

File: governance/snapshot-vote-analysis/vote_analysis.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+        votes_short_by = round(votes_needed_for_quorum - votes_for, 2)
+        print(f'The minimum number of additional votes needed to pass was {votes_short_by}')
+        if index_hodler_df['Balance'].max() > votes_short_by:
+            lowest_rank_balance_needed_to_pass =  min(index_hodler_df.index[index_hodler_df['Balance'] > votes_short_by]) + 1
+            print(f'This means the initiative could have passed if any wallet with balance ranked in the top {lowest_rank_balance_needed_to_pass} had voted')
+    voter_balance_array = vote_df['balance'].to_numpy()
+    compute_gini('All INDEX voter balances', voter_balance_array)
+    vote_df.plot.bar(y='balance', figsize=(24,8)).xaxis.set_visible(False)
+    if votes_for > votes_needed_for_quorum:
+        quorum_threshold_index = min(vote_df.index[vote_df['cumulative_dist'] > votes_needed_to_pass])
```
Comment (josephd8) [inline governance/snapshot-vote-analysis/vote_analysis.py:113]: `votes_needed_to_pass` is not a known variable

**Comment by josephd8** [comment]
@craigtutterow awesome work. I pulled down your code and got it all running. This is all a great first step. I left a few comments directly in-line on some issues I ran into.

As for next steps, I think there are two levels to this project:
1) delivering a synopsis/report on the takeaways for the Coop regarding governance and voting power - this is to help the Coop have immediate context around the current state of governance/voting power in order to make appropriate decisions to improve governance
2) building tools for continued analysis on governance and voting at the Coop (and possibly beyond the Coop)

I am not sure if I have a vision yet on what we need for #2 to drive the kind of impact needed to make the effort worthwhile, but I want to explore it further. My thoughts are that we use this foundation to really drive some immediate takeways on #1, and then we think about #2 as a more long-term project potentially.

**Review by josephd8 — CHANGES_REQUESTED** [review]

**Comment by josephd8** [comment]
@craigtutterow here are a couple of comments from the Set team on what they are hoping to learn:

- Dylan: "not sure this is in his wheel house, but i'd be curious who isn't voting. we only hit... 25% of circulating tokens or so? I'd be curious where the other 75% are? if they're whales or a long tail?"

- Greg: For each DG2 vote, what % of $INDEX cast was by whales vs. non-whales (building on: https://gov.indexcoop.com/t/iip-voting-analysis/676) . For IIP-7 it was ~80% vs. 20%

**Comment by craigtutterow** [comment]
Will look into this over the weekend and wrap up by Monday evening. I think the most generically useful and lightweight utility would be a calculator to simulate votes needed to reach quorum and approval rates. We could have an API query that fetches the token distribution, and then inputs/sliders for the various deciles. Also, we could help build the analytics/analysis into Snapshot itself, so it's available as a tab on the vote pages.

**Comment by josephd8** [comment]
Hey @craigtutterow just checking in here - any progress? I want to re-iterate that while we have tossed around some cool ideas for tooling, the immediate need here is much simpler - just looking for some answers on the questions brought up by Dylan and Greg (noted a few comments ago) as well as the original questions noted in the issue. A simple pdf report with some visuals, or something like that would suffice!

Let me know, thanks.

**Pull request merged**